virtio-devices/src/net.rs

```diff
@@ -84,6 +95,20 @@ impl EpollHelperHandler for NetCtrlEpollHandler {
                 {
                     error!("Failed to process control queue: {:?}", e);
                     return true;
+                } else {
+                    match self.queue.needs_notification() {
+                        Ok(true) => {
+                            if let Err(e) = self.signal_used_queue(self.queue_index) {
```

Comment:
Question: I am assuming this is an improvement as the original code was correct and works fine. What's the benefit of using `signal_used_queue()` after processing the control queue events?

Reply:
A related question is: what is the purpose of returning `bool` from `ctrl_q.process()` at success, e.g. `Ok(!used_desc_heads.is_empty())` (which is not being used)?

Reply:
> A related question is: what is the purpose of returning `bool` from `ctrl_q.process()` at success, e.g. `Ok(!used_desc_heads.is_empty())` (which is not being used)?

This is residual from when we didn't have the virtio-queue `VIRTIO_F_EVENT_IDX` handling and would be used to determine whether to generate an interrupt. I've removed it the last commit since as you point it is unnecessary.

Reply:
> Question: I am assuming this is an improvement as the original code was correct and works fine. What's the benefit of using `signal_used_queue()` after processing the control queue events?

It's actually required for correctness. After adding used descriptors you must tell the guest unless the guest has suppressed it using `VIRTIO_F_EVENT_IDX` (which `Queue::needs_notification()` will tell you about.) In the Linux kernel it actually drives the control queue without an interrupt based approach and instead spins on the descriptor being consumed. However we shouldn't rely on that behaviour and instead should adhere to the correct behaviour per the spec.